// @grpc/grpc-js/src/generated/google/protobuf/DescriptorProto.ts
export async function generateScript(product: Product, opts?: {
  model?: string
  systemPrompt?: string
  userTemplate?: string
}): Promise<string> {
  const startTime = Date.now()
  
  try {
    const config = getConfig()
    const apiKey = config.OPENAI_API_KEY
    if (!apiKey) {
      throw new AppError('OPENAI_API_KEY not configured', ErrorCode.MISSING_CONFIG, 500)
    }

    const model = opts?.model || config.OPENAI_MODEL || 'gpt-4o'

    const systemPrompt = opts?.systemPrompt || config.OPENAI_SYSTEM_PROMPT || `You are a direct-response product video copywriter for Nature's Way Soil.

Return ONLY valid JSON with this exact structure (no extra text):

{
  "voiceover": "the full spoken narration (75-95 words, ends with 'Visit natureswaysoil.com for more info')",
  "scenes": [
    {
      "seconds": "0-8",
      "avatarText": "short spoken line for avatar",
      "brollKeyword": "exact Pexels search phrase for beautiful B-roll (e.g. 'slow motion liquid fertilizer spraying on green lawn')",
      "visualDesc": "short description for HeyGen"
    },
    // 4-6 more scenes
  ]
}

Rules:
- First scene must be a strong 0-3 second hook
- Voiceover must be natural spoken English, farmer/soil educator tone
- brollKeyword must be short, specific, portrait-friendly Pexels search terms
- Total video ~30 seconds
- Never describe the video itself in voiceover`

    const userTemplate = opts?.userTemplate || config.OPENAI_USER_TEMPLATE || `Create a 30-second vertical video script for: {title}

Product details: {details}

Audience: homeowners, gardeners, small farms who want living-soil solutions.

Return ONLY the JSON above.`

    const title = String(product.title || product.name || product.id || '').trim()
    const details = String(product.details || product.description || product.Description || product.caption || '').trim()

    const filled = userTemplate
      .replaceAll('{title}', title)
      .replaceAll('{details}', details)

    logger.info('Generating structured script with B-roll', 'OpenAI', { productTitle: title })

    const text = await rateLimiters.execute('openai', async () => {
      return withRetry(async () => {
        const res = await axios.post(
          'https://api.openai.com/v1/chat/completions',
          {
            model,
            messages: [
              { role: 'system', content: systemPrompt },
              { role: 'user', content: filled },
            ],
            temperature: 0.7,
            max_tokens: 600,
          },
          { headers: { Authorization: `Bearer ${apiKey}` } }
        )

        return res.data?.choices?.[0]?.message?.content?.trim() || ''
      }, { maxRetries: 3 })
    })

    // ... (keep the rest of your error handling and metrics exactly as they were)

    const duration = Date.now() - startTime
    metrics.incrementCounter('openai.success')
    metrics.recordHistogram('openai.duration', duration)
    logger.info('Structured script generated', 'OpenAI', { duration })

    return text
  } catch (error: any) {
    // ... keep your existing catch block unchanged
  }
}
